abi/realloc.go

```diff
@@ -0,0 +1,52 @@
+package abi
+
+import "unsafe"
+
+// realloc allocates or reallocates memory for Component Model calls across
+// the host-guest boundary.
+//
+// Note: the use of uintptr assumes 32-bit pointers, e.g. GOOS=wasm32 when compiled for WebAssembly.
+//
+//go:export cabi_realloc
+//go:wasmexport cabi_realloc
+func realloc(ptr unsafe.Pointer, size, align, newsize uintptr) unsafe.Pointer {
+	if newsize <= size {
```

Comment:
Don't you want to check if `ptr` is not `nil`, or `uintptr(ptr) != 0`?

Reply:
If newsize is 0, and ptr is 0, then it will always return 0. This is somewhat based on the Rust implementation.